HP-77 : [관리자]모든 상품 조회 endpoint 및 커스텀 페이지 생성

DongHun3946 opened the pull request:
# 티켓
HP-77

# 설명
1. 커스텀 페이지 생성
2. [관리자 페이지] 모든 상품 조회 endpoint 생성
3. ExceptionCode 에 CATEGORY_NOT_FOUND 예외 추가

## 타입
- [ ] 버그
- [x] 작업
- [ ] 기능 향상

# 테스트 방법
단위 테스트 및 postman

# 체크리스트
- [x] 작성한 코드가 새로운 에러를 만들지 않았습니다
- [x] 코드가 코드 컨벤션에 모두 만족하는지 확인하였습니다
- [x] 작성한 코드에 관한 unit test를 작성하였고 모든 테스트를 통과하였습니다

DongHun3946 committed:
HP-77 : 모든 상품 조회 endpoint 및 커스텀 페이지 생성

jamkim1997 commented:
이거 제가 짠 코드들도 들어있는 것 같은데 PR 잘못올리신것 같아요

jamkim1997 committed:
HP-53: 메인페이지 모든 상품 조회 endpoint (#21)

* HP-53: 메인페이지 모든 상품 조회 endpoint

* HP-53: 테스트 코드 고침

jamkim1997 commented on src/main/java/com/happypill/application/repository/product/ProductRepository.java:
```diff
@@ -1,11 +1,57 @@
+    WHERE pi.language = :language
+    ORDER BY p.productId DESC
+""")
+    Page<AdminProductListResponse> getAllProductInfosByLanguage(@Param("language") Language language, Pageable pageable);
```
Repository에선 entity 만 나가는게 맞는 것 같습니다.
ProductResponse를 받으시려면 Repo 에선 Product를 받아서 서비스에서  mapping 하는게 맞는 것 같아요

jamkim1997 commented on src/main/java/com/happypill/application/repository/product/ProductRepository.java:
```diff
@@ -1,11 +1,57 @@
+    AND p.category.categoryId = :categoryId
+    ORDER BY p.productId DESC
+""")
+    Page<AdminProductListResponse> getAllProductInfosByCategoryAndLanguage(@Param("categoryId") Long categoryId, @Param("language") Language language, Pageable pageable);
```
Repository에선 entity 만 나가는게 맞는 것 같습니다.
ProductResponse를 받으시려면 Repo 에선 Product를 받아서 서비스에서  mapping 하는게 맞는 것 같아요

jamkim1997 commented on src/test/java/com/happypill/application/service/admin/AdminProductServiceTest.java:
```diff
@@ -5,28 +5,35 @@
 @SpringBootTest
-@Transactional
 @ActiveProfiles("test")
 @AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
```
지헌님 말씀대로 @ActiveProfiles("test")
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE) 
삭제

jamkim1997 commented on src/test/java/com/happypill/application/service/admin/AdminProductServiceTest.java:
```diff
@@ -46,77 +53,116 @@ class AdminProductServiceTest {
+    @DisplayName("[특정 상품 조회] ProductPrice 가 존재하지 않으면 에러가 발생한다.")
+    void getProductDetails_1() {
+        //given
+        this.productPriceRepository.deleteAllInBatch();
```
after each 로 모든 db가 초기화 되는데 혹시 여기서 또 삭제하는 이우ㅠ가 있나요 ?

DongHun3946 replied:
@BeforeEach 에서 객체 생성할 때  Category, Product, ProductInfo,, ProductPrice 가 생성되는데 이 테스트에서는 ProductPrice가 없는 경우여서 삭제했습니다

jamkim1997 commented on src/test/java/com/happypill/application/service/admin/AdminProductServiceTest.java:
```diff
@@ -46,77 +53,116 @@ class AdminProductServiceTest {
-
-        ProductPrice productPrice = ProductPrice.of(1L, 3500, true, product);
-        productPriceRepository.save(productPrice);
+        this.productInfoRepository.deleteAllInBatch();
```
after each 로 모든 db가 초기화 되는데 혹시 여기서 또 삭제하는 이우ㅠ가 있나요 ?

jamkim1997 commented on src/main/java/com/happypill/api/controller/admin/AdminProductController.java:
```diff
@@ -1,22 +1,41 @@
-    @GetMapping("/{productId}") //특정 상품 조회
+    //모든 상품 조회
+    @GetMapping
+    //@PreAuthorize("hasRole('ADMIN')")
```
이거 나중에 지헌님 코드 커밋하면 바꾸실거죠 ??
앞에 Todo 붙여주세요 ~

jamkim1997 commented on src/main/java/com/happypill/api/controller/admin/AdminProductController.java:
```diff
@@ -1,22 +1,41 @@
+
+    //특정 상품 조회
+    @GetMapping("/{productId}")
+    //@PreAuthorize("hasRole('ADMIN')")
```
이거 나중에 지헌님 코드 커밋하면 바꾸실거죠 ??
앞에 Todo 붙여주세요 ~

jiheon2234 replied:
동훈님이 SecurityConfig에 ADMIN 추가해주실수 잇나요?

jamkim1997 commented on src/main/java/com/happypill/application/service/admin/AdminProductService.java:
```diff
@@ -23,6 +30,23 @@ public class AdminProductService {
+        Language language = Language.parseLanguage(locale.getLanguage());
+
+        if(categoryId != null){
+            boolean isExist = categoryRepository.existsById(categoryId);
```
바로 getCategoryById 를 Optional<Category>로 받아서

categoryRepository.getCategoryById().orElstThrow (...) 로 한줄로 바꿀 수 있지 않나요 ?

jamkim1997 commented on src/main/java/com/happypill/application/repository/product/ProductRepository.java:
```diff
@@ -1,11 +1,57 @@
+    JOIN pi.product p
+    JOIN ProductPrice pp ON pp.product = p AND pp.isUsed = true
+    WHERE pi.language = :language
+    ORDER BY p.productId DESC
```
descending 으로 받으시는 이유가 있을까요 ?

DongHun3946 replied:
가장 최근에 등록된 상품들이 출력되게 하기 위해서입니다

DongHun3946 committed:
fix : 테스트 코드 중 불필요한 어노테이션 제거

jamkim1997 commented on src/main/java/com/happypill/application/repository/product/ProductRepository.java:
```diff
@@ -1,11 +1,57 @@
+    JOIN ProductPrice pp ON pp.product = p AND pp.isUsed = true
+    WHERE pi.language = :language
+    AND p.category.categoryId = :categoryId
+    ORDER BY p.productId DESC
```
여기도 내림차순으로 받는 이유가 있으,ㄹ까요 ?

DongHun3946 replied:
가장 최근에 등록된 상품들이 출력되게 하기 위해서입니다

jiheon2234 commented on src/main/java/com/happypill/api/controller/admin/AdminProductController.java:
```diff
@@ -1,22 +1,41 @@
 @RequestMapping("/api/admin/products")
 public class AdminProductController {
 
+    private static final String LANGUAGE_HEADER = "Language";
```
제가 이부분 다음작업으로 수정하겠습니다. LocaleResolver 사용이 맞는거같아요

jiheon2234 commented on src/test/java/com/happypill/application/service/admin/AdminProductServiceTest.java:
```diff
@@ -46,77 +51,116 @@ class AdminProductServiceTest {
+        //when //then
+        assertThatThrownBy(() -> adminProductService.getAllProducts(1000L, pageable, locale))
+                .isInstanceOf(BusinessException.class)
+                .hasMessageContaining("카테고리");
```
hasMessageContaing으로 하는건, 나중에 메시지 변경됬을때 귀찮을 것 같습니다

jiheon2234 commented on src/main/java/com/happypill/application/pagination/CustomPage.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.List;
+
+public record CustomPage<T>(
```
그냥 의견인데, InfiniteScroll 단어가 네이밍에 좋을 것 같아요 이건 나중에 한번 얘기해봐야겠네요

jiheon2234 reviewed (approved):
굿

jiheon2234 committed:
Hp-73/user email (#22)

스프링 시큐리티 ROLE_ 프리픽스 사용으로 변경
redis 및 testcontainer redis 도입
이메일 검증 코드 전송 및 확인 구현

DongHun3946 committed:
fix : Page<dto> 구조 Page<entity> 로 변경

DongHun3946 committed:
fix : rebase 후 conflict 해결

DongHun3946 committed:
Revert "fix : rebase 후 conflict 해결"

This reverts commit 855034a8415a31a4e76651a75d973050dc1f2422.
